### PR TITLE
[IMP] doc : expand the coding guidelines (2e wave !)

### PR DESCRIPTION
**Pull request by jem-odoo**
- correct typos
- clarify some points
- precise the xml notation : record vs act_window/menuitem/...
- answer the question 'how to organize wizard and report'
- ...

### CODE REVIEW

File: doc/reference/guidelines.rst [hunk 1]
```diff
@@ -4,30 +4,37 @@
 Odoo Guidelines
 =================
 
-This page introduce the new Odoo Coding Guidelines. This guidelines
+This page introduces the new Odoo Coding Guidelines. This guidelines
```
Comment (rim-odoo) [inline doc/reference/guidelines.rst]: I think the plural of "this" is "these" so I'd write "These guidelines" ;-)

File: doc/reference/guidelines.rst [hunk 2]
```diff
@@ -4,30 +4,37 @@
...
 
+This guidelines should be applied to every new module. These guidelines will be applied to old module only in case of refactoring.
+
 Module structure
 ================
 
 Directories
 ------------
-A module is organised in a few directory :
+A module is organised in a some important directories. These directories aims to contains the business core of the module ; having a look at them should make understand the purpose of the module.
```
Comment (rim-odoo) [inline doc/reference/guidelines.rst]: a some ? :D

File: doc/reference/guidelines.rst [hunk 3]
```diff
@@ -255,13 +303,13 @@ Inside these 3 groups, the imported lines are alphabetically sorted.
 Idioms
 -------
 
-- Prefer ``%`` over ``.format()``, prefer ``%(varname)`` instead of position (This is better for translation)
-- Try to avoid generators and decorators
-- Always favor *Readability* over *conciseness* or using the language features or idioms.
+- Prefer ``%`` over ``.format()`` (when only one variable to replace in a string), prefer ``%(varname)`` instead of position (when multiple variables are to be replaced in a string). This makes the translation easier for the translators community.
+- Avoid to create generators and decorators: only use the ones provide by the Odoo API.
+- Always favor *readability* over *conciseness* or using the language features or idioms.
```
Comment (rim-odoo) [inline doc/reference/guidelines.rst]: I don't catch the link between readability and language features

File: doc/reference/guidelines.rst [hunk 4]
```diff
@@ -255,13 +303,13 @@ Inside these 3 groups, the imported lines are alphabetically sorted.
...
 - Use list comprehension, dict comprehension, and basic manipulation using ``map``, ``filter``, ``sum``, ... They make the code easier to read.
 - The same applies for recordset methods : use ``filtered``, ``mapped``, ``sorted``, ...
 - Each python file should have ``# -*- coding: utf-8 -*-`` as first line
-- Use the ``UserError`` defined in ``openerp.exceptions`` instead of overriding ``Warning``, or find a more appropriate exception in *exceptions.py*
+- Use the ``UserError`` defined in ``openerp.exceptions`` instead of overriding ``Warning``, or find a more appropriate exception in *exceptions.py* (only from v9).
```
Comment (rim-odoo) [inline doc/reference/guidelines.rst]: then put this line only in master guidelines... :)

File: doc/reference/guidelines.rst [hunk 5]
```diff
@@ -392,15 +442,15 @@ Prefix your commit with
...
+  self-explanatory (long enough) including the name of the module that
   has been changed and the reason behind that change. Do not use
   single words like "bugfix" or "improvements".
 
-- Avoid commits which simultaneously impacts lots of modules. Try to
-  splits into different commits where impacted modules are different
+- Avoid commit which simultaneously impact lots of modules. Try to
+  split them into different commits where impacted modules are different
   (It will be helpful when we are going to revert that module
   separately).
```
Comment (rim-odoo) [inline doc/reference/guidelines.rst]: whut?

File: doc/reference/guidelines.rst [hunk 6]
```diff
@@ -4,30 +4,37 @@
...
 
+These guidelines should be applied to every new module. These guidelines will be applied to old module only in case of refactoring.
+
 Module structure
 ================
 
 Directories
 ------------
-A module is organised in a few directory :
+A module is organised in some important directories. These directories aim to contain the business core of the module; having a look at them should make understand the purpose of the module.
```
Comment (dreispt) [inline doc/reference/guidelines.rst]: I suggest having subdirectories to be optional for modules with few file. The obvious case would be the module extending one model and one view. In those cases we would instead end the file names with append `_model.py`or _view.xml`.

File: doc/reference/guidelines.rst [hunk 7]
```diff
@@ -141,27 +181,23 @@ Security, View and Action
 
 Use the following pattern :
 
-* For a menu : *<model_name>_menu*
+* For a menu : *<model_name>_menu_<action>* where *action* is a little explanation of the action (use *root* for the main menu).
 * For a view : *<model_name>_view_<view_type>*, where *view_type* is kanban, form, tree, search, ...
```
Comment (dreispt) [inline doc/reference/guidelines.rst]: Isn't `_view_` redundant here? `<model_name>_<view_type>` would be expressive enough.

File: doc/reference/guidelines.rst (hunk 3 above)
Comment (dreispt) [inline doc/reference/guidelines.rst]: I see the point: there are cases where very concise implementations can turn out very hard to read.

File: doc/reference/guidelines.rst [hunk 8]
```diff
@@ -255,22 +304,24 @@ Inside these 3 groups, the imported lines are alphabetically sorted.
...
-
+- Every method used to compute data for a 'stat button' should use a `read_group` or a SQL query. This aims to improve performance (by computing data in only on query)
 
 
 Symbols
 -------
-
-- Odoo Python Class : use camelcase for code in api v8, underscore lowercase notation for old api.
+- Model name (using the dot notation, prefix by the module name) :
+    - When defining an Odoo Model : use singular form of the name (`res.partner` and `sale.order` instead of `res.partnerS` and `saleS.orderS`)
```
Comment (dreispt) [inline doc/reference/guidelines.rst]: And the first word should be the "main module" it relates to, such as `account.x` or `sale.x`.

File: doc/reference/guidelines.rst [hunk 9]
```diff
@@ -392,15 +443,15 @@ Prefix your commit with
...
-- Always put meaning full commit message: commit message should be
-  self explanatory (long enough) including the name of the module that
+- Always put meaningful commit message: commit message should be
+  self-explanatory (long enough) including the name of the module that
   has been changed and the reason behind that change. Do not use
   single words like "bugfix" or "improvements".
 
-- Avoid commits which simultaneously impacts lots of modules. Try to
-  splits into different commits where impacted modules are different
+- Avoid commit which simultaneously impact lots of modules. Try to
```
Comment (dreispt) [inline doc/reference/guidelines.rst]: commits

File: doc/reference/guidelines.rst (hunk 6 above)
Comment (bealdav) [inline doc/reference/guidelines.rst]: @dreispt +1

**Comment by odony** [comment]
To add:
- [ ] Security Guidelines section
  - [ ] SQL injection guide: basic rules, alternatives for complicated meta-queries
  - [ ] Private/public methods
  - [ ] XSS prevention: `t-raw`/`t-esc`, `$.html()`
  - [ ] CSRF explanation >=9.0
  - [ ] Access rights on ICP
  - [ ] Controllers with SUPERUSER_ID
- [ ] No hardcoding of XML IDs in business code
- [ ] Translation guidelines: no merging of fragments, t-translation=off, etc.
- [ ] TODO: ....

File: doc/reference/guidelines.rst [hunk 10]
```diff
@@ -6,24 +6,31 @@
 Odoo Guidelines
 ===============
 
-This page introduce the new Odoo Coding Guidelines. These guidelines
-aim to improve the quality of the code (better readability of source,
-...) and  Odoo Apps. Indeed, proper code ought ease maintenance, aid
-debugging, lower complexity and promote reliability.
+This page introduce the new Odoo Coding Guidelines. These guidelines aim to improve the quality of the code (better readability of source, ...) and Odoo Apps. Indeed, proper code ought ease maintenance, aid debugging, lower complexity and promote reliability.
```
Comment (tde-banana-odoo) [inline doc/reference/guidelines.rst]: - introduces
- Those aim to improve ... (remove redundant guidelines)
- remove parenthesis, use i.e. or e.g., never know which one to use
- proper code eases maintenance, aids debugging, lowers complexity and promotes reliability

File: doc/reference/guidelines.rst [hunk 11]
```diff
@@ -6,24 +6,31 @@
...
+
+These guidelines should be applied to every new module, and new developpment. These guidelines will be applied to old module **only** in case of code refactoring (migration to new API, big refactoring, ...).
 
 Module structure
 ================
 
 Directories
 -----------
-A module is organised in a few directory :
+A module is organised in some important directories. These directories aim to contain the business core of the module; having a look at them should make understand the purpose of the module.
```
Comment (tde-banana-odoo) [inline doc/reference/guidelines.rst]: - remove "some"
- Those contain the business logic ...
- Just looking at them should help understanding the module purpose.

File: doc/reference/guidelines.rst [hunk 12]
```diff
@@ -6,24 +6,31 @@
...
 
 - *data/* : demo and data xml
 - *models/* : models definition
-- *controllers/* : contains controllers (http routes).
+- *controllers/* : contains controllers (HTTP routes).
 - *views/* : contains the views and templates
 - *static/* : contains the web assets, separated into *css/, js/, img/, lib/, ...*
 
+Other directories compose the module.
```
Comment (tde-banana-odoo) [inline doc/reference/guidelines.rst]: Other optional directories

File: doc/reference/guidelines.rst [hunk 13]
```diff
@@ -6,24 +6,31 @@
...
+
+- *data/* : contains the data (in XML form)
```
Comment (tde-banana-odoo) [inline doc/reference/guidelines.rst]: Already present above

File: doc/reference/guidelines.rst [hunk 14]
```diff
@@ -6,24 +6,31 @@
...
+- *wizard/* : regroup the transient models (formerly *osv_memory*) and their views.
```
Comment (tde-banana-odoo) [inline doc/reference/guidelines.rst]: regroups

File: doc/reference/guidelines.rst [hunk 15]
```diff
@@ -44,16 +51,36 @@ For instance, *sale* module introduces ``sale_order`` and
 ``<main_model>`` files will be named :file:`models/sale_order.py` and
 :file:`views/sale_order_views.py`.
 
-
 For *data*, split them by purpose : demo or data. The filename will be
 the main_model name, suffixed by *_demo.xml* or *_data.xml*.
 
-For *controller*, the only file should be named *main.py*.
+For *controllers*, the only file should be named *main.py*. Otherwise, if you need to inherit an existing controller from another module, its name will be *<module_name>.py*. Unlike *models*, each controller should be contained in a separated file.
```
Comment (tde-banana-odoo) [inline doc/reference/guidelines.rst]: Each controller ? Each class maybe ?

File: doc/reference/guidelines.rst [hunk 16]
```diff
@@ -44,16 +51,36 @@ For instance, *sale* module introduces ``sale_order`` and
...
+
+For *static files*, since the resources can be used in different contexts (frontend, backend, both), they will be included in only one bundle. So, CSS/Less, JavaScript and XML files should be suffixed with the name of the bundle type. i.e.: *im_chat_common.css*, *im_chat_common.js* for 'assets_common' bundle, and *im_chat_backend.css*, *im_chat_backend.js* for 'assets_backend' bundle.
```
Comment (tde-banana-odoo) [inline doc/reference/guidelines.rst]: - Regarding static files
- resources can be used in different contexts (..) and are included in only one bundle (-> use the present)
- remove the "So", you french pig http://franceisnice.com/wp-content/uploads/2015/04/tumblr_nkpelk2xIo1u0k6deo2_2501.gif

File: doc/reference/guidelines.rst [hunk 17]
```diff
@@ -44,16 +51,36 @@ For instance, *sale* module introduces ``sale_order`` and
...
+For modules having only one file, the convention will be *<module_name>.ext* (i.e.: *project.js*).
```
Comment (tde-banana-odoo) [inline doc/reference/guidelines.rst]: If the module owns only one file, use <module_name> ...

File: doc/reference/guidelines.rst [hunk 18]
```diff
@@ -44,16 +51,36 @@ For instance, *sale* module introduces ``sale_order`` and
...
+Don't link data (image, libraries) outside Odoo: don't use an
```
Comment (tde-banana-odoo) [inline doc/reference/guidelines.rst]: Use do not, not don't.

File: doc/reference/guidelines.rst [hunk 19]
```diff
@@ -44,16 +51,36 @@ For instance, *sale* module introduces ``sale_order`` and
...
+URL to an image but copy it in our codebase instead.
+
+For *data*, split them by purpose: data or demo. The filename will be
```
Comment (tde-banana-odoo) [inline doc/reference/guidelines.rst]: Regarding data

File: doc/reference/guidelines.rst [hunk 20]
```diff
@@ -44,16 +51,36 @@ For instance, *sale* module introduces ``sale_order`` and
...
+the *main_model* name, suffixed by *_data.xml* or *_demo.xml*.
+
+For *wizards*, the naming convention is :
```
Comment (tde-banana-odoo) [inline doc/reference/guidelines.rst]: Regarding + remove unnecessary "the"

File: doc/reference/guidelines.rst [hunk 21]
```diff
@@ -134,6 +172,16 @@ When declaring a record in XML,
         </field>
     </record>
 
+Some syntax equivalences exists, and can be used:
```
Comment (tde-banana-odoo) [inline doc/reference/guidelines.rst]: Odoo supports custom tags acting as syntax sugar

File: doc/reference/guidelines.rst [hunk 22]
```diff
@@ -266,22 +325,27 @@ Inside these 3 groups, the imported lines are alphabetically sorted.
...
-- Try to avoid generators and decorators
-- Always favor *Readability* over *conciseness* or using the language features or idioms.
+- Prefer ``%`` over ``.format()`` (when only one variable to replace in a string), prefer ``%(varname)`` instead of position (when multiple variables are to be replaced in a string). This makes the translation easier for the translators community.
+- Avoid to create generators and decorators: only use the ones provide by the Odoo API.
+- Always favor *readability* over *conciseness* or using the language features or idioms.
 - Use list comprehension, dict comprehension, and basic manipulation using ``map``, ``filter``, ``sum``, ... They make the code easier to read.
 - The same applies for recordset methods : use ``filtered``, ``mapped``, ``sorted``, ...
-- Each python file should have ``# -*- coding: utf-8 -*-`` as first line
-- Use the ``UserError`` defined in ``openerp.exceptions`` instead of overriding ``Warning``, or find a more appropriate exception in *exceptions.py*
+- Each python file should have ``# -*- coding: utf-8 -*-`` as first line.
```
Comment (tde-banana-odoo) [inline doc/reference/guidelines.rst]: Actually should not be necessary?

File: doc/reference/guidelines.rst [hunk 23]
```diff
@@ -266,22 +325,27 @@ Inside these 3 groups, the imported lines are alphabetically sorted.
...
 - Document your code (docstring on methods, simple comments for the tricky part of the code)
-- Use meaningful variable/class/method names
-
+- Use meaningful variable/class/method names.
+- Every method used to compute data for a 'stat button' should use a ``read_group`` or a SQL query. This aims to improve performance (by computing data in only on query).
```
Comment (tde-banana-odoo) [inline doc/reference/guidelines.rst]: ... when possible. Remove the parenthesis.

File: doc/reference/guidelines.rst [hunk 24]
```diff
@@ -263,25 +321,380 @@ Inside these 3 groups, the imported lines are alphabetically sorted.
...
+
+- Each python file should have ``# -*- coding: utf-8 -*-`` as first line.
+- Always favor *readability* over *conciseness* or using the language features or idioms.
+- Don't use ``.clone()``
+
+.. code-block:: python
+
+    # bad
+    new_dict = my_dict.clone()
+    new_list = old_list.clone()
```
Comment (guewen) [inline doc/reference/guidelines.rst]: Of course that's bad as nor `[].clone()` nor `{}.clone()` actually exist :)
I think you mean `{}.copy()` but `[].copy()` does not exist neither. Not sure why you'd say `{}.copy()` is bad though.

**Pull request closed without merging**